feature/experimental/tunnel/otg_tests/tunnel_interface_based_resize_test/README.md

```diff
@@ -0,0 +1,97 @@
+D[DUT:Port3\lag1] --> E[Port3\lag1:OTG];
+f[DUT:Port4\lag2] --> G[Port4\lag2:OTG];
+H[DUT:Port5\lag2] --> I[Port5\lag2:OTG];
+J[DUT:Port6] --> K[Port6:OTG];
+```
+
+
+
+## Subtest-1: Setup Tunnels as per requirements
+1. The test should configure 32 GRE tunnel interfaces per address family and each should be configured with different /32 and /128 tunnel destination prefixes of the same /27 and /64 prefix range. Each of these tunnel interfaces should use the loopback address as the source-address for their respective address-families. Hence, we will have total 64xtunnel interfaces, 32 per address-family as source and destination addresses of the tunnels.
+2. Add static routes to provide reachability for the /27 IPv4 and /64 IPv6 tunnel destination prefixes in bullet#1 above to point at lag1, lag2 and port6 for ECMP.
+3. Configure dualstack IPv4 and an IPv6 addresses for each of the 64xtunnel interfaces. The IP addresses can be of /30 and /127 prefix-lengths.
+4. Pick 2 different /24 IPv4 and /64 IPv6 subnets each to emulate destination of payload traffic to be tunneled. Note: This is for the destination prefixes of the flows generated by OTG:Port1.
+5. Add 128 different v4 and v6 static routes (64 for each address-family) for the 2 sets of /24 IPv4 and /64 IPv6 payload prefixes in bullet#4 above. The static routes should be such that, the first set of 1x /24 IPv4 and 1x /64 IPv6 prefixes should point at the 32x tunnel interfaces configured w/ IPv4 tunnel destination and the second set of 1x /24 IPv4 and 1x /64 IPv6 prefixes should point at the 32x tunnel interfaces w/ IPv6 tunnel destinations.
+
+### Verification:
+ *   Ensure all tunnel interfaces are up
+ *   Ensure all LAG and singleton interfaces are up
+ *   Run Health-1.1. checks
+
+## Subtest-2: Gather baseline stats by passing traffic
+ *   Start with the configuration in Subtest-1 above.
+ *   The test should send 2000 IPv4 and 2000 1Pv6 flows from the ATE:Port1 to the DUT:Port1. These flows should equally be split between the 2x /24 IPv4 and 2x /64 IPv6 destination prefixes mentioned in Subtest-1 bullet#4 above. The flows should use variations in the following,
+        * Source IP address
+        * Source Port
+        * Destination IP adresses.
+        * Destination Port
+        * Payload protocol
+ *   Incoming traffic on DUT:PORT1 should be load shared across all 64 Tunnel interfaces for encapsulation and then routed out the combination of lag and singleton interfaces, i.e. lag1, lag2 and Port6.
+ *   Follow all the "verifications" steps in [*] below
+
+## Subtest-3: Reduce number of tunnel interfaces 
+ *   Start with the state in Subtest-2 above with active traffic flow.
+ *   Reduce the number of Tunnel interfaces, e.g. from 32 to 16 for each of the tunnel destination group by address-family:
+        * If the static routes are used to forward traffic to tunnel, please disable or delete the static routes in this test to simulate the reduction in available paths.
```

Comment:
Shouldn't the tunnel endpoints be removed by deleting ` interfaces/interfaces/interface/tunnel/config/dest/ ` entries?  If there are unreachable tunnel endpoints (due to no route), is the DUT behavior well defined?     Is the intent to test for this condition?  Or just the removal of endpoints in the config?

Reply:
Well, we will have multiple static routes for the same destination prefixes to be encaped, mapped to several next hops which basically will be the tunnel IP addresses. If you remove some static routes, the remaining static routes should still be functional. Net effect is that we will have reduced number of tunnel interfaces. 
Other approach is as you said, remove the tunnel destination addresses from the tunnel interfaces. I am guessing that will take down the tunnel interfaces and hopefully will make the corresponding static routes passive and therefore not useable. But this approach has several nuances
1. We need to test if the chain of things I mentioned above really happens so
2. I dont know how different NOS will take the change. Possible that a NOS wouldn't let us commit a config because the tunnel interface has missing tunnel destination.

Tl;dr: Earlier approach seems more predictable.

Reply:
Ah, I see.  I agree that removing only a portion of the static routes has a simple, expected behavior.    It's confusing to call this 'removal of tunnel interfaces'.  This approach is 'removal of static routes', causing some  the tunnel interfaces to not be used.  

This IMO is testing if adding/removing static routes causes packet loss.  This is not testing if adding and removing tunnels causes packet loss.   

Changing the static routes is an ok test and I don't think we have coverage of this in our current tests that use static routes (which seem to be [present in gribi tests, while the static route feature has none](https://github.com/search?q=repo%3Aopenconfig%2Ffeatureprofiles%20static%20route&type=code)).   But I think this portion of the test is not really exercising tunnels, it's static routing that is getting tested and would apply equally to tunnel interfaces and physical interfaces.

Do you mind updating the description to say "removal of static routes, causing some  the tunnel interfaces to not be used?"

I think we need a subtest in this readme which covers actually changing the tunnel configuration, not just static routes.    Here's why:  I think what we want to happen operationally is configuration generation will create a new test of tunnels and this will be applied using a config replace at /interfaces in one step.   There won't be a multi-step configuration change where routing changes (static or otherwise) is applied to 'drain' or 'isolate' a device and then make the configuration change.    So we need to test this way.     Example high level logic:
* Configure tunnel interface and routes set 1 in single gnmi.Replace
* Start traffic
* Configure tunnel interface and routes set 2 in single gnmi.Replace
* Check for packet loss

In production, are we using static routes?  Or dynamic routing?  Maybe the routes for this subtest should be isis+bgp?

Reply:
`Do you mind updating the description to say "removal of static routes, causing some the tunnel interfaces to not be used?"`
- Done

`I think we need a subtest in this readme which covers actually changing the tunnel configuration, not just static routes. `
- Done

`In production, are we using static routes? Or dynamic routing? Maybe the routes for this subtest should be isis+bgp?`
- Static routes. We will learn about the destination prefixes for which the traffic has to be encaped over IBGP from the WAN ports. But then these static routes are in place to change the local router's behavior to punt the traffic to the tunnel interface instead using static routes. These routes will be locally significant for these routers.